lib/timers.js

```diff
@@ -62,55 +119,72 @@ function listOnTimeout() {
   var now = Timer.now();
   debug('now: %s', now);
 
-  var diff, first, threw;
-  while (first = L.peek(list)) {
-    diff = now - first._idleStart;
+  var diff, timer, threw;
+  while (timer = L.peek(list)) {
+    diff = now - timer._idleStart;
+
+    // Check if this loop iteration is too early for the timer.
+    // This happens if one of two things is true:
+    // - There are more timers scheduled for later in the list. (Most common)
+    // - The earliest timer was canceled (unenrolled). (Less common)
     if (diff < msecs) {
       list.start(msecs - diff, 0);
       debug('%d list wait because diff is %d', msecs, diff);
       return;
-    } else {
-      L.remove(first);
-      assert(first !== L.peek(list));
+    }
+
+    // The actual logic for when a timeout happens.
+
+    L.remove(timer);
+    assert(timer !== L.peek(list));
 
-      if (!first._onTimeout) continue;
+    if (!timer._onTimeout) continue;
+
+    var domain = timer.domain;
+    if (domain) {
 
       // v0.4 compatibility: if the timer callback throws and the
```

Comment:
This comment is obsolete, and doesn't describe the current state of the code base. It seems that this PR is a good opportunity to remove it.